File: 3. Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        mapping = {}
        maxAns = 0
        for i in range(len(s)):
            mapping.clear()
            
            mapping[s[i]] = 1
            
            j = i + 1
            while j < len(s):
                
                if s[j] in mapping:
                    mapping[s[j]] = mapping[s[j]] + 1
                else:
                    mapping[s[j]] = 1
                
                if mapping[s[j]] == 2:
                    break;
                j = j + 1
            if (j - i) > maxAns:
                maxAns = (j - i)
            
           
        return maxAns
        
 class Solution(object):
    def lengthOfLongestSubstring(self, s):
        mapping = {}
        
        left = -1
        maxAns = 0
        for i in range(len(s)):
            
            if s[i] in mapping and mapping[s[i]] > left:
                left = mapping[s[i]]
            
            mapping[s[i]] = i
           
            maxAns = max(maxAns, i - left)
            
        
        return maxAns
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        mapping = [-1] * 256
        
        left = -1
        maxAns = 0
        for i in range(len(s)):
            
            if mapping[(ord)s[i]] > left:
                left = mapping[(ord)s[i]]
            
            mapping[(ord)s[i]] = i
           
            maxAns = max(maxAns, i - left)
            
        
        return maxAns
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        mapping = set()
        
        left = 0
        maxAns = 0
        i = 0
        while i < len(s):
            
            if s[i] not in mapping:
                
                mapping.add(s[i])
                i = i + 1
                maxAns = max(maxAns, len(mapping))
            else:
                mapping.remove(s[left])
                left = left + 1
            
            
            
        
        return maxAns
 

File: 3. Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> mapping = new HashMap<Character, Integer>();
        int maxAns = 0;
        int i, j;
        for(i = 0; i < s.length(); i++){
            mapping.clear();
            mapping.put(s.charAt(i), 1);
            for(j = i + 1; j < s.length(); j++){
                if(mapping.containsKey(s.charAt(j))){
                    mapping.put(s.charAt(j), mapping.get(s.charAt(j)) + 1);
                } else {
                    mapping.put(s.charAt(j), 1);
                }
                
                if(mapping.get(s.charAt(j)) == 2){
                    break;
                }
                
            }
            if((j - i) > maxAns){
                maxAns = (j - i);
            }
            
        }
        
           
        return maxAns;
    }
}
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> mapping = new HashMap<Character, Integer>();

        
          
        //滑动窗口左边位置 它总在滑动窗口最左边字符的前一位，以便计算窗口长度，初始化为-1
        int left = -1;
        int maxAns = 0;
        for(int i = 0; i < s.length(); i++){
            
            if(mapping.containsKey(s.charAt(i)) && mapping.get(s.charAt(i)) > left){
                left = mapping.get(s.charAt(i));
            }
            mapping.put(s.charAt(i), i);//储存字符的位置无论是否出现过
            //更新最大长度
            maxAns = Math.max(maxAns, i - left);
            
        }
        return maxAns;
    }
}
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] mapping = new int[256];
        
        Arrays.fill(mapping, -1);
        
          
        //滑动窗口左边位置 它总在滑动窗口最左边字符的前一位，以便计算窗口长度，初始化为-1
        int left = -1;
        int maxAns = 0;
        for(int i = 0; i < s.length(); i++){
            
            if(mapping[s.charAt(i)] > left){
                left = mapping[s.charAt(i)];
            }
            mapping[s.charAt(i)] = i;//储存字符的位置无论是否出现过
            //更新最大长度
            maxAns = Math.max(maxAns, i - left);
            
        }
        return maxAns;
    }
}
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> mapping = new HashSet<Character>();
        
        int left = 0;
        int maxAns = 0;
        int i = 0;
        while(i < s.length()){
            
            if(!mapping.contains(s.charAt(i))){
                
                mapping.add(s.charAt(i++));
                maxAns = Math.max(maxAns, mapping.size());
            } else {
                mapping.remove(s.charAt(left++));
            }
            
            
            
        }
        return maxAns;
        
    }
}
